---
title: "Oral Microbiome and Mucositis Severity in Cancer Patients"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Oral Microbiome and Mucositis Severity in Cancer Patients}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r knitr_options, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    warning = FALSE,
    message = FALSE
)
```

Mucositis is a painful and common side effect of many cancer treatments for head
and neck cancers. Recently, Zhang et al. (2023) uncovered links between the
progression of mucositis and the composition of the oral microbiome. They
studied a cohort of more than 150 patients over time and collected 16S rRNA
sequencing data to understand how microbiome community profiles relate to
mucositis severity.  We will revisit these data in this vignette. The dataset is
available in the accompanying ExperimentData package, `CancerOralMicrobiome`,
which can be downloaded and installed from Bioconductor using
`BiocManager::install("CancerOralMicrobiome")`.

In the first part, we will analyze the data at the sample level. We will compute
a summary of mucositis severity across all patients and use this to organize the
bars in a phylobar plot. This will allow us to draw conclusions at both the
individual OTU and higher community levels, reflecting the subtle effects
reported by Zhang et al. In the second part, we will focus on temporal trends.
Rather than visualizing the composition for individual patients, we will compute
smoothed time series relative to the date of cancer treatment initiation,
stratified by mucositis severity. This will allow us to identify patterns in
microbiome dynamics that appear to be associated with mucositis progression.

# Setup

The block below loads the relevant packages. We use the tidyverse packages for
preprocessing.

```{r setup}
library(ape)
library(dplyr)
library(phylobar)
library(tibble)
library(tidyr)
```

Next, we load the data, which is organized as a phyloseq object with slots for
the phylogenetic tree, sample metadata, and sample-level abundances.

```{r read_in_data}
f <- tempfile()
download.file("zenodo.org/records/17179980/files/zhang2023-ps.rds", f) # will be replaced by ExperimentHub package, CancerOralMicrobiome
ps <- readRDS(f)
```

# Sample-Level Analysis

We now begin the sample-level analysis. The original dataset includes 1103
samples. To simplify visualization, we subset to 100 representative samples
chosen from a clustering with $K = 100$, as implemented by `subset_cluster`. In
the subsetted data, rows correspond to representative samples and columns
correspond to OTU abundances. We do not filter any OTUs,

```{r filter_and_subset}
keep_ix <- rowMeans(otu_table(ps) > 0) > 0.01
ps <- prune_taxa(taxa_names(ps)[keep_ix], ps)
x_mat <- t(otu_table(ps))
x_mat <- subset_cluster(x_mat, k = 100)
```

One issue is that two OTUs now never appear among the visualized samples.  Let's
add a small pseudocount to make sure each leaf node has at least some mass.

```{r normalize}
x_mat[1, which(colSums(x_mat) == 0)] <- .00001
x_mat <- x_mat / rowSums(x_mat)
```

## Sample order

By default,  phylobar orders samples in the stacked bar plot according to sample
similarity. While this creates clean visualizations, it fails to address our
primary interest: how variation in community composition relates to mucositis.
Therefore, we instead order the samples by the mucositis severity of the
patients from which they were gathered. The block below computes patient-level
averages of mucositis severity and then organizes samples first by patient
severity and, within each patient, by time relative to treatment initiation.

```{r sample_order}
sample_order <- data.frame(sample_data(ps)) |>
    rownames_to_column("sample") |>
    group_by(id) |>
    mutate(mean_mucositis = mean(mucositis)) |>
    ungroup() |>
    arrange(mean_mucositis, timediff) |>
    pull(sample)
```

Although the ordering includes every sample in the dataset, our plots use only
the representative subset from the clustering step. We finalize the ordering by
subsetting the sample order accordingly.

```{r compute_order_indices}
row_ix <- sample_order[sample_order %in% rownames(x_mat)]
col_ix <- hclust(dist(t(x_mat)))$order
```

## Visualization

The reorganized matrix is then used as input to phylobar. Samples to the right
in the stacked bar have more severe mucositis on average. We use the argument
`sample_show_fall = FALSE` to use the custom ordering defined above. The
`sample_label_space = 75` argument gives a bit of extra vertical space for the
sample labels, which are a bit longer than usual.

```{r plot_sample_phylobar}
phylobar(
    x_mat[row_ix, col_ix],
    phy_tree(ps),
    width = 900,
    hclust_order = FALSE,
    sample_font_size = 6,
    sample_label_space = 75
)
```

## Interpretation

We highlight a few interesting findings using screenshots from the interactive
plots. For example, the tallest bar on the far left is one of 16 (out of 1103)
samples that has more than 10 reads of OTU 200, and it is in fact dominated by
this OTU 200. This suggests possible contamination or quality-control issues.

<!-- Replaced Markdown image with HTML figure to keep source lines short -->
<figure>
  <img src="https://i.imgur.com/fuIRxOv.png"
       alt="Questionable OTU dominating a sample" />
  <figcaption>
    A questionable OTU that seems to dominate one of the samples.
  </figcaption>
</figure>

Next, we highlight two phylogenetic subtrees to explore associations with
mucositis severity. In the subtree labeled in purple (descendants of node
0.859.2), higher abundance appears to be associated with lower severity -- the
purples bars are larger on the left hand side of the stacked barplot. In
contrast, in the green subtree (descendants of node 0.948.4), higher abundances
are associated with higher severity.

<figure>
  <img src="https://i.imgur.com/SVPR6fY.png"
       alt="High-level phylogenetic groups linked to mucositis" />
  <figcaption>
    A few high-level phylogenetic groups that seem to be associated with
    mucositis severity.
  </figcaption>
</figure>

Next, we collapse a few subtrees, resulting in a smaller tree layout and letting
us focus on some of the remaining phylogenetic groups. We then highlight a bar
(red) with large values in one of the severe mucositis patients. THe associated
strain, OTU0910, shows high abundance in a relatively small number of patients,
and those patients tend to experience worse than average mucositis.

<figure>
  <img src="https://i.imgur.com/9mYiMBg.png"
       alt="Collapsed tree focusing on a isolated leaf." />
  <figcaption>
    By collapsing parts of the tree, we can study patterns in a few isolated
    leaves without losing context of the entire tree.
  </figcaption>
</figure>

# Temporal Analysis

The example above visualizes stacked bars for each sample. However, we might be
more interested in temporal trends averaged across all samples—for example, how
microbiome community structures change over the course of treatment among people
with more vs. less severe mucositis. In the next part, we'll create a
streamgraph alternative to the stacked bars above, where the stream is defined
by time relative to start of treatment.

## Streamgraph helpers

The helper function below computes a kernel smoothing of abundances for each
species over time. We consider windows of size two days and take the average
across all patients within that window. A final Gaussian kernel averaging is
applied to further smooth across adjacent windows.

```{r smooth_time_series}
smooth_time_series <- function(df, time_col = "timediff", pseudocount = 1, round_degree = 2, bandwidth = 5) {
    # Round timediffs
    df <- df |>
        mutate(time_bin = round_degree * round(.data[[time_col]] / round_degree)) |>
        arrange(time_bin) |>
        group_by(time_bin) |>
        summarise(
            value = (sum(value) + pseudocount) / (n() + pseudocount)
        )

    # Fill gaps with 0
    tibble(time_bin = seq(min(df$time_bin), max(df$time_bin), by = round_degree)) |>
        left_join(df, by = "time_bin") |>
        mutate(value = replace_na(value, 0)) |>
        reframe(
            smooth_value = ksmooth(time_bin, value, "normal", bandwidth, n.points = n())$y,
            time_bin = ksmooth(time_bin, value, "normal", bandwidth, n.points = n())$x
        )
}
```

The block below converts the original phyloseq object into a pair of data
frames, since this format is easier to manipulate with dplyr. We also compute
the average mucositis score for each individual, identified by the ID column.
This derived feature is useful for understanding disease severity. An
alternative analysis could also consider latent class grouping, as discussed in
Zhang et al. (2023).

```{r convert_to_dataframes}
x_mat <- t(otu_table(ps)) |>
    data.frame() |>
    rownames_to_column("sample")
metadata <- sample_data(ps) |>
    data.frame() |>
    rownames_to_column("sample") |>
    group_by(id) |>
    mutate(mean_mucositis = mean(mucositis))
```

## Input data preparation

We now reshape the data into long-format so that we can apply the smoothing
function separately across each species (`feature` in the data below) and
disease severity level. Several patients have very late follow-up measurements,
but we focus only on those within 55 days of treatment. The result, `x_smooth`,
applies the kernel smoothing separately to mroe vs. less severe mucositis cases,
then combines the results into a single matrix—first grouped by severity, then
sorted by time. This allows us to study the two trends within a single stacked
bar plot.

```{r prepare_smoothing_input}
x_long_format <- x_mat |>
    left_join(metadata, by = "sample") |>
    filter(between(timediff, -5, 55)) |>
    select(id, timediff, mean_mucositis, starts_with("OTU")) |>
    pivot_longer(starts_with("O"), names_to = "feature", values_to = "value")

x_smooth <- x_long_format |>
    group_by(feature, severity = if_else(mean_mucositis > 3.5, "S", "NS")) |>
    group_modify(~ smooth_time_series(.x)) |>
    unite(row, time_bin, severity, sep = "") |>
    pivot_wider(names_from = feature, values_from = smooth_value) |>
    column_to_rownames("row")
x_smooth <- x_smooth / rowSums(x_smooth)
```

## Visualization

We can now create our phylobar visualization, using the same phylogenetic tree
as above. The x-axis labels indicate severity and time window. All of the more
severe cases are shown on the left side of the figure.

```{r plot_timeseries_phylobar}
phylobar(
    x_smooth,
    phy_tree(ps),
    width = 900,
    hclust_order = FALSE
)
```

## Interpretation

Browsing across different phylogenetic groups, we see that the descendants of
0.538 are more abundant at the early time points in patients with more severe
mucositis. In both the higher and lower severity groups, however, the abundance
of this subtree decreases over time. In contrast, OTU0911 remains roughly stable
across both groups. In the final few time windows for the lower severity group,
it appears to become slightly more abundant, though this may be an artifact of
averaging over the smaller number of samples available at 55 days after
treatment start.

<figure>
  <img src="https://i.imgur.com/bQmh0IV.png"
       alt="Phylobar streamgraph overview of taxonomic trajectories" />
  <figcaption>
    An overview phylobar streamgraph, showing a few high-level taxonomic
    groups with differential trajectories across high and low severity groups.
  </figcaption>
</figure>

Interacting next with the stacked bars, we can identify several individual OTUs
of interest. For example, OTU0307 (highlighted in purple) becomes more abundant
at later time points, especially among patients with low mucositis. This could
represent a candidate species for designing a therapy aimed at managing
mucositis. OTU0297 (green) appears briefly midway through treatment and into
recovery among low-severity cases, illustrating the transient effects that can
arise in microbiome data. Finally, OTU0943 appears slightly more abundant among
those with lower severity, though the effect is tentative.

<figure>
  <img src="https://i.imgur.com/iIcbMs0.png"
       alt="Zoomed phylobar streamgraph highlighting OTUs of interest" />
  <figcaption>
    A zoomed in phylobar streamgraph, focusing on some OTUs of interest.
  </figcaption>
</figure>

These observations emerged after only a few minutes of interacting with the
phylobar visualization. This highlights the exploratory potential of the
approach and shows how the same dataset can be viewed at multiple phylogenetic
resolutions to uncover biologically meaningful trends.

# Session Info

```{r session_info}
sessionInfo()
```