---
title: "Diet Data"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{mouse_data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    message = FALSE,
    warning = FALSE
)
```

Diet is a key factor shaping the gut microbiome. In a recent mouse study, 
researchers profiled gut microbial communities under different dietary conditions 
to examine how nutrient intake influences community structure. The dataset, available
as a phyloseq object, includes microbial feature tables, taxonomic annotations, 
and sample metadata.

In this vignette, we use these data to demonstrate how phylobar can produce 
phylogenetically informed visualizations of microbiome composition. We begin 
by normalizing feature abundances and reconstructing the taxonomic tree. We 
then generate a phylobar plot to highlight community shifts across diet groups, 
with a focus on clades where dietary effects are most pronounced.

```{r load-libs}
library(phylobar)
library(phyloseq)
library(DESeq2)
library(MicrobiomeStat)
```

# Setup
We begin by loading the mouse diet study data, which is stored remotely and 
can be loaded into R as a phyloseq object.

```{r download-data}
dietpath <- tempfile(fileext = ".rda")
download.file("https://go.wisc.edu/l68i2z", dietpath, mode = "wb")
diet <- get(load(dietpath))

diet
```

To prepare the data for visualization, we first convert the phyloseq object
into a MicrobiomeStat object using mStat_convert_phyloseq_to_data_obj().
This enables us to apply DESeq2 normalization, which corrects for sequencing
depth differences across samples. We then transpose the normalized feature
matrix so that rows correspond to taxa and columns to samples, as expected by 
phylobar().

```{r preprocess}
diet_mstat <- mStat_convert_phyloseq_to_data_obj(diet)
diet_mstat <- mStat_normalize_data(diet_mstat, "DESeq")
x <- t(diet_mstat$data.obj.norm$feature.tab)
```

Finally, we construct a taxonomy-based tree. The taxonomic annotations are
prefixed (p_, f_, etc.) to conform with standard formats and then
converted into a tree structure using taxonomy_to_tree(). This tree will serve
as the backbone for our visualization.

```{r build-tree}
taxa <- tax_table(diet)
taxa[, 1] <- paste0("p_", taxa[, 1])
taxa[, 2] <- paste0("f_", taxa[, 2])
taxa <- cbind(Kingdom = "k_Bacteria", taxa)
tree <- taxonomy_to_tree(taxa)
```

At this point, we have two aligned objects ready for visualization: a normalized 
abundance matrix (x) and a taxonomy-derived tree (tree). We now pass these objects 
to phylobar(). The resulting plot shows stacked bar charts of taxa abundances, 
aligned with the hierarchical tree structure.

```{r render-phylobar}
phylobar(x, tree, width=800)
```

# Interpretation

We showcase several notable patterns using screenshots from the interactive plots: most 
mice display a co-dominance of Firmicutes (purple) and Bacteroidetes (teal), but 
sample_52 shows a clear Firmicutes dominance, underscoring how dietary interventions
can shift the balance between these two phyla.

![](https://imgur.com/a/soZPl5i)

Next, we observe the abundance of the Clostridium cluster IV family acros samples.
The variability in its abundance suggests that while Clostridium cluster IV is 
widespread, its contribution to community structure differs across individuals or conditions.

![](https://imgur.com/a/TGBiIg3)


# Session Info

```{r session-info}
sessionInfo()
```