<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Interactive construction of stacked barplots using hierarchies • phylobar</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Interactive construction of stacked barplots using hierarchies">
<meta name="description" content='The phylobar package supports interactive visualization of microbiome data by allowing a stacked barplot to be constructed using a guiding taxonomic or phylogenetic hierarchy. The package provides a strategy and collapsing and expanding the hierarchy to different levels of resolution and then for interactively "painting" the stacked barplot by placing the mouse over different subtrees. This makes it possible to interactively test different color palettes at different resolution and identify taxonomic groups with interesting variation before settling on a final stacked barplot. One advantage of the approach is that multiple levels of taxonomic resolution can be compared at once within the same view.'>
<meta property="og:description" content='The phylobar package supports interactive visualization of microbiome data by allowing a stacked barplot to be constructed using a guiding taxonomic or phylogenetic hierarchy. The package provides a strategy and collapsing and expanding the hierarchy to different levels of resolution and then for interactively "painting" the stacked barplot by placing the mouse over different subtrees. This makes it possible to interactively test different color palettes at different resolution and identify taxonomic groups with interesting variation before settling on a final stacked barplot. One advantage of the approach is that multiple levels of taxonomic resolution can be compared at once within the same view.'>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">phylobar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/atlas.html">Tipping Points in the Atlas 1006 Dataset</a></li>
    <li><a class="dropdown-item" href="articles/cancer_oral_microbiome.html">Oral Microbiome and Mucositis Severity in Cancer Patients</a></li>
    <li><a class="dropdown-item" href="articles/customizing_files.html">Customizing Style in Phylobar</a></li>
    <li><a class="dropdown-item" href="articles/diet_analysis.html">Differences in the Community Composition due to Diet</a></li>
    <li><a class="dropdown-item" href="articles/exporting.html">Exporting to Vector Graphics Format</a></li>
    <li><a class="dropdown-item" href="articles/global_patterns.html">Working with phyloseq and TreeSummarizedExperiment versions of the Global Patterns Dataset</a></li>
    <li><a class="dropdown-item" href="articles/hfhs.html">The effect of a high-fat high-sugar (HFHS) diet on the mouse microbiome</a></li>
    <li><a class="dropdown-item" href="articles/taxonomies.html">Building Trees from Taxonomies</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mkdiro-O/phylobar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="phylobar">phylobar<a class="anchor" aria-label="anchor" href="#phylobar"></a>
</h1></div>
<p>phylobar is a visualization package for constructing stacked barplots by interactively “painting” an associated tree. This is an alternative to defining a color palette at a fixed taxonomic resolution. It also helps avoid the issue of grouping all rare taxa into a color for “other” — you can selectively highlight specific rare taxa while leaving others at their background color.</p>
<p>The main function, phylobar, takes two inputs: a table of potentially normalized counts and a tree structure which can come from either a phylogeny or a table of taxonomic assignments. The vignettes show how to convert alternative data formats (e.g., phyloseq) into this input format. You can export interactive snapshots to SVG using the SVG crowbar package, as shown in the “Exporting Views” vignette.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of phylobar using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_packages</span><span class="op">(</span><span class="st">"mkdir-o/phylobar"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>This block applies phylobar to a small random data set:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mkdiro-O/phylobar" class="external-link">phylobar</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html" class="external-link">rtree</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fl">20</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fl">100</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/phylobar.html">phylobar</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://raw.githubusercontent.com/krisrs1128/LSLab/main/assets/img/rtree_recording.gif"></p>
<p>GitHub doesn’t support interactive blocks in READMEs, so we’ve included a recording above. You can interact with other example output in our precompiled <a href="https://mkdiro-o.github.io/phylobar/articles" class="external-link">articles</a>. Click a node to collapse a subtree, press the control key to introduce a new color, and press the escape key to freeze the view.</p>
</div>
<div class="section level2">
<h2 id="common-tasks">Common Tasks<a class="anchor" aria-label="anchor" href="#common-tasks"></a>
</h2>
<ul>
<li>
<a href="https://mkdiro-o.github.io/phylobar/articles/customizing_files.html" class="external-link">Customizing style</a>: Phylobar supports a few styling customizations, like changing the size of the text labels or the color palette. This vignette walks through currently available options.</li>
<li>
<a href="https://mkdiro-o.github.io/phylobar/articles/taxonomies.html" class="external-link">Building a tree from a taxonomy</a>: Rather than interacting with abstract phylogenetic trees it can be helpful to choose stacked bar color using taxonomic assignments. This vignette gets a quick overview of constructing trees from taxonomy tables, including checks to make sure that the input is formatted properly.</li>
<li>
<a href="https://mkdiro-o.github.io/phylobar/articles/exporting.html" class="external-link">Exporting views</a>: While interactivity is useful for exploration, we often need to export a specific static view to discuss with others. This vignette gives an alternative to simple screenshots that preserves image quality and supports, editing and software like illustrator or inkscape.</li>
<li>
<a href="https://mkdiro-o.github.io/phylobar/reference/subset_cluster.html" class="external-link">Subsetting to representative samples</a>: Stacked bar plots can be cumbersome when there are many samples (e.g., &gt; 1000) present, because the bars become too thin. This vignette includes some helper functions to create views from representative samples.</li>
</ul>
</div>
<div class="section level2">
<h2 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h2>
<p>The full function reference can be found here: <a href="https://mkdiro-o.github.io/phylobar/reference/index.html" class="external-link uri">https://mkdiro-o.github.io/phylobar/reference/index.html</a>.</p>
</div>
<div class="section level2">
<h2 id="learn-more">Learn More<a class="anchor" aria-label="anchor" href="#learn-more"></a>
</h2>
<p>Additional vignettes provide fully developed examples for quality control, cross group comparison, and longitudinal analysis. The package design follows the focus-plus-context principle in data visualization – our paper provides an in-depth discussion. The package is implemented using the htmlwidgets package for R-JavaScript integration. The JavaScript code is available on npm as phylobar-js.</p>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>You can reach us by creating an <a href="(https://github.com/mkdiro-O/phylobar/issues)" class="external-link">Issue</a> or emailing <a href="mailto:ksankaran@wisc.edu" class="email">ksankaran@wisc.edu</a>. We appreciate your interest and will respond promptly.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/mkdiro-O/phylobar/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/mkdiro-O/phylobar/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>CC0</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing phylobar</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Kris Sankaran <br><small class="roles"> Author, maintainer, funder </small> <a href="https://orcid.org/0000-0002-9415-1971" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Megan Kuo <br><small class="roles"> Author </small>   </li>
<li>Saritha Kodikara <br><small class="roles"> Author </small>   </li>
<li>Jiadong Mao <br><small class="roles"> Author </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kris Sankaran, Megan Kuo, Saritha Kodikara, Jiadong Mao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
