<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Building Trees from Taxonomies • phylobar</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Building Trees from Taxonomies">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">phylobar</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/atlas.html">Tipping Points in the Atlas 1006 Dataset</a></li>
    <li><a class="dropdown-item" href="../articles/covid_immunology.html">Cell-type Hierarchy Informs COVID-19 Immunology</a></li>
    <li><a class="dropdown-item" href="../articles/customizing_files.html">Customizing Style in phylobar</a></li>
    <li><a class="dropdown-item" href="../articles/diet_analysis.html">Differences in the Community Composition due to Diet</a></li>
    <li><a class="dropdown-item" href="../articles/exporting.html">Exporting to Vector Graphics Format</a></li>
    <li><a class="dropdown-item" href="../articles/global_patterns.html">Working with phyloseq and TreeSummarizedExperiment versions of the Global Patterns Dataset</a></li>
    <li><a class="dropdown-item" href="../articles/hfhs.html">The effect of a high-fat high-sugar (HFHS) diet on the mouse microbiome</a></li>
    <li><a class="dropdown-item" href="../articles/runtime_evaluation.html">Runtime Evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/taxonomies.html">Building Trees from Taxonomies</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mkdiro-O/phylobar/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Building Trees from Taxonomies</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mkdiro-O/phylobar/blob/HEAD/vignettes/taxonomies.Rmd" class="external-link"><code>vignettes/taxonomies.Rmd</code></a></small>
      <div class="d-none name"><code>taxonomies.Rmd</code></div>
    </div>

    
    
<p>Taxonomy provides useful context for interpreting microbiome data,
and taxonomy tables are often provided as output from microbiome
sequencing pipelines. However, phylobar requires a phylo object as
input. This is a more formal representation of a tree than a table of
taxonomy assignments. To support the conversion, the package includes a
helper function to convert such assignments into the required tree
format. Since taxonomy tables lack a standardized representation, small
discrepancies can produce invalid trees. This article collects examples
from other vignettes of the common issues that can arise when
constructing a taxonomy tree and goes into more depth about the
strategies that were used to address them.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/emmanuelparadis/ape" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mkdiro-O/phylobar" class="external-link">phylobar</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="checking-that-that-the-induced-tree-is-correct">Checking that that the induced tree is correct<a class="anchor" aria-label="anchor" href="#checking-that-that-the-induced-tree-is-correct"></a>
</h2>
<p>Before looking at potential issues, let’s review how to check whether
an input tree is valid. In the block below, we generate a random tree
using the rtree function from ape. We can confirm validity with the
<code>checkValidPhylo</code> function. If the output is TRUE, the tree
can be used for phylobar visualizations. If not, the function provides
hints on how to adjust the taxonomy table so that
<code>taxonomy_to_tree</code> produces a valid tree.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html" class="external-link">rtree</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 20 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 19 </span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="introducing-a-missing-root-node">Introducing a missing root node<a class="anchor" aria-label="anchor" href="#introducing-a-missing-root-node"></a>
</h2>
<p>One common issue is the absence of a root node that links all
descendant microbes. Many taxonomy tables implicitly assume the Bacteria
kingdom and begin at the phylum level. In such cases, validity checks
may fail with an error like the one below. This occurs because each
phylum is treated as a separate tree, rather than being connected into a
single tree with the Kingdom as the root.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">atlas1006</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">atlas1006</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 128 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 23 </span></span>
<span><span class="co">#&gt;   FATAL: nodes and tips should appear only once in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt;   FATAL: the root node should not appear in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>This problem can be resolved by introducing a new “kingdom” column
from which all phyla descend. With this fix,
<code>taxonomy_to_tree</code> correctly builds edges between the first
and second columns. Re-running the validity check now produces a valid
tree object, which is used in the earlier Atlas vignette.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">atlas1006</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>Kingdom <span class="op">=</span> <span class="st">"Bacteria"</span>, <span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_prefix.html">add_prefix</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 130 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 31 </span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="skipping-over-missing-taxonomic-assignments">Skipping over missing taxonomic assignments<a class="anchor" aria-label="anchor" href="#skipping-over-missing-taxonomic-assignments"></a>
</h2>
<p>Another common issue involves missing taxonomic assignments. phylobar
can skip missing entries, but only if they are encoded as NA. If missing
values are stored as character strings (e.g., “unclassified”), they are
treated as valid categories. Since they often appear at multiple levels
of resolution and under different parents, this breaks the tree
structure.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HFHSdata</span>, package <span class="op">=</span> <span class="st">"LUPINE"</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="va">HFHSdata</span><span class="op">$</span><span class="va">filtered_taxonomy</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 8 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 287 </span></span>
<span><span class="co">#&gt;   FATAL: each tip must appear once in 'edge'</span></span>
<span><span class="co">#&gt;   FATAL: all nodes should appear at least twice in 'edge'</span></span>
<span><span class="co">#&gt;   MODERATE: some nodes are of degree 1 or less</span></span>
<span><span class="co">#&gt;   FATAL: nodes and tips should appear only once in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>To avoid this, missing values should be explicitly coded as NA. This
approach is illustrated in our high-fat, high-sugar diet vignette, where
the taxonomy table is pre-processed to replace placeholder strings with
NA. Here is the taxonomy before any correction. Notice that the NAs are
not properly coded – we see the prefix coming from the taxonomic level,
but no corresponding name.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="co">#&gt;            X1          X2                X3              X4                X5</span></span>
<span><span class="co">#&gt; OTU_13 OTU_13 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_21 OTU_21 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_7   OTU_7 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_29 OTU_29 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_30 OTU_30 k__Bacteria     p__Firmicutes   c__Clostridia  o__Clostridiales</span></span>
<span><span class="co">#&gt; OTU_14 OTU_14 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt;                       X6   X7   X8</span></span>
<span><span class="co">#&gt; OTU_13          f__S24-7  g__  s__</span></span>
<span><span class="co">#&gt; OTU_21          f__S24-7  g__  s__</span></span>
<span><span class="co">#&gt; OTU_7           f__S24-7  g__  s__</span></span>
<span><span class="co">#&gt; OTU_29  f__Rikenellaceae  g__  s__</span></span>
<span><span class="co">#&gt; OTU_30               f__  g__  s__</span></span>
<span><span class="co">#&gt; OTU_14          f__S24-7  g__  s__</span></span></code></pre></div>
<p>We address this in the block below, checking for whether the name
ended with <code>__</code> and replacing those with explicit NA
value.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="va">taxa</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">X1</span>, <span class="va">X1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html" class="external-link">across</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span>, <span class="op">~</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_starts.html" class="external-link">str_ends</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"_"</span><span class="op">)</span>, <span class="cn">NA</span>, <span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 X2                X3              X4                X5</span></span>
<span><span class="co">#&gt; OTU_13 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_21 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_7  k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_29 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt; OTU_30 k__Bacteria     p__Firmicutes   c__Clostridia  o__Clostridiales</span></span>
<span><span class="co">#&gt; OTU_14 k__Bacteria  p__Bacteroidetes  c__Bacteroidia  o__Bacteroidales</span></span>
<span><span class="co">#&gt;                       X6   X7   X8     X1</span></span>
<span><span class="co">#&gt; OTU_13          f__S24-7 &lt;NA&gt; &lt;NA&gt; OTU_13</span></span>
<span><span class="co">#&gt; OTU_21          f__S24-7 &lt;NA&gt; &lt;NA&gt; OTU_21</span></span>
<span><span class="co">#&gt; OTU_7           f__S24-7 &lt;NA&gt; &lt;NA&gt;  OTU_7</span></span>
<span><span class="co">#&gt; OTU_29  f__Rikenellaceae &lt;NA&gt; &lt;NA&gt; OTU_29</span></span>
<span><span class="co">#&gt; OTU_30              &lt;NA&gt; &lt;NA&gt; &lt;NA&gt; OTU_30</span></span>
<span><span class="co">#&gt; OTU_14          f__S24-7 &lt;NA&gt; &lt;NA&gt; OTU_14</span></span></code></pre></div>
<p>Once this transformation is made, the table can be used to construct
a valid tree.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 212 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 80 </span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="avoiding-duplicated-names-across-different-taxonomic-levels">Avoiding duplicated names across different taxonomic levels<a class="anchor" aria-label="anchor" href="#avoiding-duplicated-names-across-different-taxonomic-levels"></a>
</h2>
<p>Another common problem arises when taxonomic names are not explicitly
distinguished across different levels of resolution. For example, in the
<code>dietswap</code> dataset, some phylum- and family-level assignments
share the same names. If uncorrected, <code>taxonomy_to_tree</code>
interprets these edges as loops, which produces an invalid tree. The
validity check will fail in this case. To see this, let’s first extract
the taxonomy table.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dietswap"</span>, package <span class="op">=</span> <span class="st">"microbiome"</span><span class="op">)</span></span>
<span><span class="va">diet_temp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">dietswap</span>, <span class="va">timepoint</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">diet</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">diet_temp</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">diet_temp</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">diet</span><span class="op">)</span></span></code></pre></div>
<p>Note the repeated phylum and family names. This causes the check to
fail.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [6 taxa by 3 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;                              Phylum            Family                    </span></span>
<span><span class="co">#&gt; Actinomycetaceae             "Actinobacteria"  "Actinobacteria"          </span></span>
<span><span class="co">#&gt; Aeromonas                    "Proteobacteria"  "Proteobacteria"          </span></span>
<span><span class="co">#&gt; Akkermansia                  "Verrucomicrobia" "Verrucomicrobia"         </span></span>
<span><span class="co">#&gt; Alcaligenes faecalis et rel. "Proteobacteria"  "Proteobacteria"          </span></span>
<span><span class="co">#&gt; Allistipes et rel.           "Bacteroidetes"   "Bacteroidetes"           </span></span>
<span><span class="co">#&gt; Anaerostipes caccae et rel.  "Firmicutes"      "Clostridium cluster XIVa"</span></span>
<span><span class="co">#&gt;                              Genus                         </span></span>
<span><span class="co">#&gt; Actinomycetaceae             "Actinomycetaceae"            </span></span>
<span><span class="co">#&gt; Aeromonas                    "Aeromonas"                   </span></span>
<span><span class="co">#&gt; Akkermansia                  "Akkermansia"                 </span></span>
<span><span class="co">#&gt; Alcaligenes faecalis et rel. "Alcaligenes faecalis et rel."</span></span>
<span><span class="co">#&gt; Allistipes et rel.           "Allistipes et rel."          </span></span>
<span><span class="co">#&gt; Anaerostipes caccae et rel.  "Anaerostipes caccae et rel."</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 117 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 22 </span></span>
<span><span class="co">#&gt;   FATAL: nodes and tips should appear only once in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt;   FATAL: the root node should not appear in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>To address this, we can add a small prefix that encodes the taxonomic
rank of each assignment. The helper function <code>add_prefix</code>
supports this concatenation. At this stage, however, running the
validity check still raises an error. As in the Atlas example, the issue
is the absence of a root node connecting the different phyla. Adding a
“Kingdom” column resolves this by linking the trees together.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_prefix.html">add_prefix</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>Kingdom <span class="op">=</span> <span class="st">"k_Bacteria"</span>, <span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 119 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 30 </span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>The issue of duplicated taxonomy names also appears in the Global
Patterns dataset. For instance, several phylum- and class-level
assignments share identical names.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, package <span class="op">=</span> <span class="st">"phyloseq"</span><span class="op">)</span></span>
<span><span class="va">chlamydiae</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_taxa-methods.html" class="external-link">subset_taxa</a></span><span class="op">(</span><span class="va">GlobalPatterns</span>, <span class="va">Phylum</span> <span class="op">==</span> <span class="st">"Chlamydiae"</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">chlamydiae</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="co">#&gt; Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:</span></span>
<span><span class="co">#&gt;        Kingdom    Phylum       Class        Order          Family             </span></span>
<span><span class="co">#&gt; 100535 "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Waddliaceae"      </span></span>
<span><span class="co">#&gt; 2936   "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Waddliaceae"      </span></span>
<span><span class="co">#&gt; 24341  "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Parachlamydiaceae"</span></span>
<span><span class="co">#&gt; 579085 "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Parachlamydiaceae"</span></span>
<span><span class="co">#&gt; 547579 "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Parachlamydiaceae"</span></span>
<span><span class="co">#&gt; 136933 "Bacteria" "Chlamydiae" "Chlamydiae" "Chlamydiales" "Parachlamydiaceae"</span></span>
<span><span class="co">#&gt;        Genus                      Species                              </span></span>
<span><span class="co">#&gt; 100535 "Waddlia"                  NA                                   </span></span>
<span><span class="co">#&gt; 2936   "Waddlia"                  "Waddliachondrophila"                </span></span>
<span><span class="co">#&gt; 24341  NA                         NA                                   </span></span>
<span><span class="co">#&gt; 579085 NA                         NA                                   </span></span>
<span><span class="co">#&gt; 547579 "CandidatusProtochlamydia" NA                                   </span></span>
<span><span class="co">#&gt; 136933 "CandidatusProtochlamydia" "CandidatusProtochlamydiaamoebophila"</span></span></code></pre></div>
<p>This duplication results in an invalid phylo.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 5 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 9 </span></span>
<span><span class="co">#&gt;   FATAL: all nodes should appear at least twice in 'edge'</span></span>
<span><span class="co">#&gt;   FATAL: nodes and tips should appear only once in the 2nd column of 'edge'</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>Again, we can resolve this using the <code>add_prefix</code>
function. This dataset raises one additional challenge: the ASV names
are stored only as row names, not as an explicit column in the taxonomy
table. Without this, we cannot reach the leaf nodes of the tree. The
solution is to introduce a new column containing the ASV
identifiers.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_prefix.html">add_prefix</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="va">taxa</span><span class="op">$</span><span class="va">ASV</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="co">#&gt;           Kingdom       Phylum        Class          Order              Family</span></span>
<span><span class="co">#&gt; 100535 K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales       F_Waddliaceae</span></span>
<span><span class="co">#&gt; 2936   K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales       F_Waddliaceae</span></span>
<span><span class="co">#&gt; 24341  K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales F_Parachlamydiaceae</span></span>
<span><span class="co">#&gt; 579085 K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales F_Parachlamydiaceae</span></span>
<span><span class="co">#&gt; 547579 K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales F_Parachlamydiaceae</span></span>
<span><span class="co">#&gt; 136933 K_Bacteria P_Chlamydiae C_Chlamydiae O_Chlamydiales F_Parachlamydiaceae</span></span>
<span><span class="co">#&gt;                             Genus                             Species    ASV</span></span>
<span><span class="co">#&gt; 100535                  G_Waddlia                                &lt;NA&gt; 100535</span></span>
<span><span class="co">#&gt; 2936                    G_Waddlia                 Waddliachondrophila   2936</span></span>
<span><span class="co">#&gt; 24341                        &lt;NA&gt;                                &lt;NA&gt;  24341</span></span>
<span><span class="co">#&gt; 579085                       &lt;NA&gt;                                &lt;NA&gt; 579085</span></span>
<span><span class="co">#&gt; 547579 G_CandidatusProtochlamydia                                &lt;NA&gt; 547579</span></span>
<span><span class="co">#&gt; 136933 G_CandidatusProtochlamydia CandidatusProtochlamydiaamoebophila 136933</span></span></code></pre></div>
<p>After this adjustment, checking the validity of the resulting tree
still returns an error. In this case, the error can be safely ignored:
it occurs when a tree contains nodes with more than two descendants.
phylobar accommodates such multifurcations without issue.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxonomy_to_tree.html">taxonomy_to_tree</a></span><span class="op">(</span><span class="va">taxa</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/checkValidPhylo.html" class="external-link">checkValidPhylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="co">#&gt; Starting checking the validity of tree...</span></span>
<span><span class="co">#&gt; Found number of tips: n = 21 </span></span>
<span><span class="co">#&gt; Found number of nodes: m = 15 </span></span>
<span><span class="co">#&gt;   FATAL: all nodes should appear at least twice in 'edge'</span></span>
<span><span class="co">#&gt; Done.</span></span></code></pre></div>
<p>To verify, we can simply plot the phylo object directly before
creating a phylobar visualization. You can check that this a static
version of the same tree that we work with in the main Global Patterns
vignette.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<p><img src="taxonomies_files/figure-html/globalpatterns-plot-1.png" alt="" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sequoia 15.7.3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/Chicago</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] stringr_1.6.0     phylobar_0.99.0   phangorn_2.12.1   purrr_1.2.1      </span></span>
<span><span class="co">#&gt; [5] htmlwidgets_1.6.4 phyloseq_1.53.0   dplyr_1.1.4       ape_5.8-1        </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] fastmatch_1.1-6     gtable_0.3.6        xfun_0.55          </span></span>
<span><span class="co">#&gt;  [4] bslib_0.9.0         ggplot2_4.0.1       rhdf5_2.53.5       </span></span>
<span><span class="co">#&gt;  [7] Biobase_2.69.1      lattice_0.22-7      quadprog_1.5-8     </span></span>
<span><span class="co">#&gt; [10] rhdf5filters_1.21.0 vctrs_0.7.0         tools_4.5.2        </span></span>
<span><span class="co">#&gt; [13] generics_0.1.4      biomformat_1.37.0   stats4_4.5.2       </span></span>
<span><span class="co">#&gt; [16] parallel_4.5.2      tibble_3.3.1        cluster_2.1.8.1    </span></span>
<span><span class="co">#&gt; [19] pkgconfig_2.0.3     Matrix_1.7-4        data.table_1.18.0  </span></span>
<span><span class="co">#&gt; [22] RColorBrewer_1.1-3  S7_0.2.1            desc_1.4.3         </span></span>
<span><span class="co">#&gt; [25] S4Vectors_0.47.4    lifecycle_1.0.5     compiler_4.5.2     </span></span>
<span><span class="co">#&gt; [28] farver_2.1.2        textshaping_1.0.4   Biostrings_2.77.2  </span></span>
<span><span class="co">#&gt; [31] Seqinfo_0.99.2      codetools_0.2-20    permute_0.9-8      </span></span>
<span><span class="co">#&gt; [34] htmltools_0.5.9     sass_0.4.10         yaml_2.3.12        </span></span>
<span><span class="co">#&gt; [37] pillar_1.11.1       pkgdown_2.1.3       crayon_1.5.3       </span></span>
<span><span class="co">#&gt; [40] jquerylib_0.1.4     MASS_7.3-65         cachem_1.1.0       </span></span>
<span><span class="co">#&gt; [43] vegan_2.7-2         iterators_1.0.14    foreach_1.5.2      </span></span>
<span><span class="co">#&gt; [46] nlme_3.1-168        tidyselect_1.2.1    digest_0.6.39      </span></span>
<span><span class="co">#&gt; [49] stringi_1.8.7       reshape2_1.4.5      splines_4.5.2      </span></span>
<span><span class="co">#&gt; [52] ade4_1.7-23         fastmap_1.2.0       grid_4.5.2         </span></span>
<span><span class="co">#&gt; [55] cli_3.6.5           magrittr_2.0.4      survival_3.8-3     </span></span>
<span><span class="co">#&gt; [58] dichromat_2.0-0.1   withr_3.0.2         scales_1.4.0       </span></span>
<span><span class="co">#&gt; [61] rmarkdown_2.30      XVector_0.49.1      multtest_2.65.0    </span></span>
<span><span class="co">#&gt; [64] igraph_2.2.1        otel_0.2.0          ragg_1.5.0         </span></span>
<span><span class="co">#&gt; [67] evaluate_1.0.5      knitr_1.51          IRanges_2.43.5     </span></span>
<span><span class="co">#&gt; [70] mgcv_1.9-3          rlang_1.1.7         Rcpp_1.1.1         </span></span>
<span><span class="co">#&gt; [73] glue_1.8.0          BiocGenerics_0.55.1 jsonlite_2.0.0     </span></span>
<span><span class="co">#&gt; [76] R6_2.6.1            Rhdf5lib_1.31.0     plyr_1.8.9         </span></span>
<span><span class="co">#&gt; [79] systemfonts_1.3.1   fs_1.6.6</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Kris Sankaran, Megan Kuo, Saritha Kodikara, Jiadong Mao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
